src/bin/pg_autoctl/service_keeper.c

```diff
@@ -350,11 +372,55 @@ keeper_node_active_loop(Keeper *keeper, pid_t start_pid)
 			continue;
 		}
 
+		/*
+		 * Upon receiving SIGTERM, start maintenance on the monitor, possibly
+		 * triggering a failover when the primary node is being shut down.
+		 */
+		if (asked_to_stop)
+		{
+			if (shutdownSequenceInProgress && couldStartMaintenanceAtShutdown)
+			{
+				log_trace("Shutdown in progress");
+			}
+			else if (keeperState->current_role != MAINTENANCE_STATE &&
+					 keeperState->current_role != PREPARE_MAINTENANCE_STATE)
+			{
+				int nodeId = keeper->state.current_node_id;
+				bool mayRetry = false;
+
+				log_info("Enabling maintenance on this node for shutdown");
+
+				/*
+				 * Make sure that at restart, we get out of maintenance, even
+				 * if we later receive a stronger signal such as SIGINT
+				 * (asked_to_stop_fast) or SIGQUIT (asked_to_quit).
+				 */
+				keeperState->service_state = SERVICE_SHUTDOWN_IN_MAINTENANCE;
+
+				if (keeper_store_state(keeper) &&
+					monitor_start_maintenance(monitor, nodeId, &mayRetry))
+				{
+					couldStartMaintenanceAtShutdown = true;
+				}
+				else
+				{
+					/* warning: we're going to try again */
+					log_warn("Failed to start maintenance "
+							 "upon receiving SIGTERM");
+				}
+			}
```

Comment:
Do we need an else here that sets `keeperState->service_state = SERVICE_SHUTDOWNING`?

Reply:
Maybe related:
When shutting down a node that is already in maintenance mode, SIGTERM is not shutting it down quickly:
```
16:17:26 31705 INFO  pg_autoctl received signal SIGTERM, terminating
16:17:26 31705 INFO  Waiting for subprocesses to terminate.
16:17:31 31705 INFO  pg_autoctl services are still running, signaling them with SIGTERM.
16:17:31 31710 INFO  Postgres controller service received signal SIGTERM, terminating
16:17:31 31705 INFO  pg_autoctl received signal SIGTERM, terminating
16:17:36 31705 INFO  pg_autoctl services are still running, signaling them with SIGINT.
16:17:36 31705 INFO  Service leader with pid 31711 has terminated, now stopping other services
16:17:36 31705 INFO  Stop pg_autoctl
```

Reply:
Even after the current fixes this issue is still present. If you enable maintenance on a node manually and then run `pg_autoctl stop` (or send SIGTERM). The process does not actually stop:
```
12:58:26 14138 INFO  FSM transition from "wait_maintenance" to "maintenance": Suspending standby for manual maintenance.
12:58:26 14138 INFO  Transition complete: current state is now "maintenance"
12:58:26 14138 INFO  New state for node 1 "node1" (localhost:5501): primary ➜ primary
12:58:35 14135 INFO  pg_autoctl received signal SIGTERM, terminating
12:58:35 14135 INFO  Waiting for subprocesses to terminate.
12:58:40 14135 INFO  pg_autoctl services are still running, signaling them with SIGTERM.
12:58:40 14137 INFO  Postgres controller service received signal SIGTERM, terminating
12:58:40 14135 INFO  pg_autoctl received signal SIGTERM, terminating
12:58:45 14135 INFO  pg_autoctl services are still running, signaling them with SIGINT.
12:58:45 14135 INFO  Service leader with pid 14138 has terminated, now stopping other services
12:58:45 14135 INFO  Stop pg_autoctl
```

On a related note, do we maybe need to increase the amount of time after which we send a SIGINT? I can imagine going to maintenance will take longer than 10 seconds.

Reply:
Fixed now. About the 10 seconds, I'm not too sure about what would be a proper value... any input? should we reuse `timeout.listen_notifications_timeout` or introduce a new one there?

Reply:
It turns out that sharing the timeout values from the keeper config isn't easy, and we can't change the value on reload anyway, so I have hard-coded a 30s timeout for now.